atox/src/main/kotlin/ui/user_profile/UserProfileFragment.kt

```diff
@@ -155,23 +170,101 @@ class UserProfileFragment : BaseFragment<FragmentUserProfileBinding>(FragmentUse
+                        flags = Intent.FLAG_GRANT_READ_URI_PERMISSION
+                    }
+                    startActivity(Intent.createChooser(shareIntent, getString(R.string.tox_id_share)))
+                }
+            }
+            .setNegativeButton(getString(R.string.close), null)
             .create()
     }
+
+    /*
+     * Function will save the image with the input bitmap and name to the cache directory as a png format.
+     * Then it will return the Uri of the image.
+     */
+    private fun saveImageForSharing(context: Context, image: Bitmap, name: String): Uri {
+        val imagesFolder = File(context.cacheDir, "shared_images")
+        val uri: Uri
+
+        imagesFolder.mkdirs()
+        val file = File(imagesFolder, "$name.png")
+        FileOutputStream(file).use { stream ->
+            image.compress(Bitmap.CompressFormat.PNG, 90, stream)
+            stream.flush()
+        }
+        uri = FileProvider.getUriForFile(context, "${BuildConfig.APPLICATION_ID}.fileprovider", file)
+
+        return uri
+    }
+
+    // Function will create and return the QR code bitmap of the Tox ID.
+    private fun createQrCode(
+        qrSize: Size = Px(
+            (
+                min(
+                    resources.displayMetrics.widthPixels,
+                    resources.displayMetrics.heightPixels
+                ) * QR_CODE_TO_SCREEN_RATIO
+                ).toInt()
+        ),
+        padding: Size = qrCodePadding
+    ): Bitmap {
+        // Creating the QR bitmap
+        val qrData = QrCode.encodeText("tox:%s".format(vm.toxId.string()), QrCode.Ecc.LOW)
+        val qrCodeColor: Int = Color.BLACK
+        val backgroundColor: Int = Color.WHITE
+        var bmpQr: Bitmap = Bitmap.createBitmap(qrData.size, qrData.size, Bitmap.Config.ARGB_8888)
+        bmpQr.setHasAlpha(true)
+        for (x in 0 until qrData.size) {
+            for (y in 0 until qrData.size) {
+                bmpQr.setPixel(x, y, if (qrData.getModule(x, y)) qrCodeColor else backgroundColor)
+            }
+        }
+        // Scaling the QR bitmap to be half of the screen's width
+        val qrSizePx = qrSize.asPx(resources).px
+        bmpQr = bmpQr.scale(qrSizePx, qrSizePx, false)
+
+        // Adding a padding to the QR bitmap
+        val paddingPx = padding.asPx(resources).px
+        val bmpQrWithPadding =
+            Bitmap.createBitmap(
+                bmpQr.width + 2 * paddingPx,
+                bmpQr.height + 2 * paddingPx,
+                Bitmap.Config.ARGB_8888
+            )
+        val canvas = Canvas(bmpQrWithPadding)
+        canvas.drawPaint(
+            Paint().apply {
+                color = backgroundColor
+                style = Paint.Style.FILL
+            }
+        )
+        canvas.drawBitmap(bmpQr, paddingPx.toFloat(), paddingPx.toFloat(), null)
+
+        return bmpQrWithPadding
+    }
+
+    private suspend fun getQrForSharing(name: String): Uri =
```

Comment:
Since this always should have the same name, I think it'd be nicer to drop the arg and inline the filename on the line  `saveImageForSharing` is called.